### ISSUE TITLE
Guru meditation right after booting 

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#596]
[DCD-BBS.adf.zip](https://github.com/dirkwhoffmann/vAmiga/files/7487901/DCD-BBS.adf.zip)

UAE:

<img width="721" alt="Bildschirmfoto 2021-11-05 um 18 33 15" src="https://user-images.githubusercontent.com/12561945/140554090-b4f60322-4995-4434-a989-bceb67a8c6d8.png">

vAmiga: 🙈

<img width="905" alt="Bildschirmfoto 2021-11-05 um 18 36 03" src="https://user-images.githubusercontent.com/12561945/140554219-b8961df0-c947-4032-836d-284493bfc33a.png">

(reported by @chris70c)

**Comment by dirkwhoffmann** [comment]
First findings:

- Works in SAE with lowest compatibility settings (immediate Blitter, turbo drive).
- Guru is due to stepping over an illegal opcode.

    [219] (167, 88) 04E0BA  0 -CB-D- 602C 1000 CPU:131 XFILES: illegalOpcodeException(45e2)
    [568] ( 70,137) FC05FA  0 -CB-D- 202C 3020 CPU:85 XFILES: RESET instruction

**Comment by mras0** [comment]
You may want to first look the at c/killborder command that's run by the startup-sequence. Trying that in @chris70c's vAmiga web port (labelled 1.1 Beta 5 2021-11-02) gives this result:
![image](https://user-images.githubusercontent.com/9952017/140663805-218479e2-e27f-4a07-81ca-ddf603bdd2ca.png)

It should look like this:
![image](https://user-images.githubusercontent.com/9952017/140663850-efb47dc4-8cea-4bcf-aab1-4ab09a703369.png)

Quick "resourcing" of the code (since it's quite short) for reference (beware of transcription errors as I did this semi-manually):

```
EntryPoint
        MOVEM.L	D0-A6, -(A7)
	MOVEA.L	AbsExecBase, A6
	LEA.L	IntuitionName(PC), A1
	JSR	_LVOOldOpenLibrary(A6)
	TST.L	D0
	BEQ.B	Out
	MOVEM.L	D0/A6, -(A7)
	MOVEA.L	D0, A6
	MOVEA.L	ActiveWindow(A6), A0
	MOVE.L	A0, -(A7)
	MOVEA.L	FirstGadget(A0), A1
	MOVEQ.L	#$ff, D0
	JSR	_LVORemoveGList(A6)
	MOVEA.L	(A7)+, A0
	MOVE.L	#$00000000, BorderLeft(A0)
	MOVE.L	#$00010800, Flags(A0)
	MOVEQ.L	#$00, D0
	MOVE.L	#$00000100, D1
	SUB.W	Height(A0), D1
	JSR	_LVOSizeWindow(A6)
	JSR	_LVORefreshWindowFrame(A6)
	MOVEM.L	(A7)+, A1/A6
	JSR	_LVOCloseLibrary(A6)
	LEA.L	DosName(PC), A1
	MOVEQ.L	#$00, D0
	JSR	_LVOOpenLibrary(A6)
	MOVEA.L	D0, A6
	JSR	_LVOOutput(A6)
	MOVE.L	#Message, D2
	MOVE.L	#$00000032, D3
	JSR	_LVOWrite(A6)
	MOVEA.L	A6, A1
	MOVEA.L	AbsExecBase, A6
	JSR	_LVOCloseLibrary(A6)
Out
	MOVEM.L	(A7)+, D0-A6
	MOVEQ.L	#$00, D0
	RTS
Message
	dc.b	$1b, 'c', $0a, ' KillBorder V2.3 ', $a9, ' Electron / DECA'
	dc.b	'DE ', $27, '93     ', $0a, $0a
IntuitionName
	dc.b	'intuition.library', $00
DosName
	dc.b	'dos.library', $00, $00, $00
```

with KS1.3 512K chip+512K slow the code will probably be loaded at $00c086f8

**Comment by dirkwhoffmann** [comment]
I've noticed that it runs in vAmiga until 0.9.10.1:

<img width="1026" alt="Bildschirmfoto 2021-11-08 um 06 51 47" src="https://user-images.githubusercontent.com/12561945/140690923-669546f9-5098-4f8b-9297-065fdb3334e6.png">

It broke in v0.9.11. Next step is to find the faulty commit.

**Comment by dirkwhoffmann** [comment]
Failed with commit https://github.com/dirkwhoffmann/vAmiga/commit/214292e01c57e76e74d3ac48552761a79a09d0b1

**Comment by dirkwhoffmann** [comment]
The demo seems to rely on certain initial values in the CPU registers. Replacing 

    for(int i = 0; i < 8; i++) reg.d[i] = reg.a[i] = 0xFFFFFFFF;

by 

    for(int i = 0; i < 8; i++) reg.d[i] = reg.a[i] = 0;

fixes the issue.

**Comment by dirkwhoffmann** [comment]
The faulty commit mentions a test called `Init/cpu/registersX`. This test hasn't found its way into the test repo. What it does is to compare the register values with the ones I've seen on my real machine. The code below shows which values were measured. However, I don't know how meaningful this test is, because a) some registers are used by Kickstart and won't contain their initial values when the program runs, and b) I don't know if these values are the same on every machine. Some degree of randomness may be involved. 

```
	; Test the initial values of all CPU registers
checkd0:
	cmp.l  #$00000000,d0
	beq    checkd1
	bra    skip
checkd1:
	cmp.l  #$80000000,d1
	beq    checkd2
	bra    skip
checkd2:
	cmp.l  #$00000001,d2
	beq    checkd3
	bra    skip
checkd3:
	cmp.l  #$00000000,d3
	beq    checkd4
	bra    skip
checkd4:
	cmp.l  #$0,d4
	beq    checkd5
	bra    skip
checkd5:	
	cmp.l  #$FFFF7FFF,d5
	beq    checkd6
	bra    skip
checkd6:
	cmp.l  #$FFFFFFFF,d6
	beq    checkd7
	bra    skip
checkd7:
	cmp.l  #$FFFFFFFF,d7
	beq    checka0
	bra    skip
checka0:
	cmpa.l #$00C0152A,a0
	beq    checka1
	bra    skip
checka1:
	cmpa.l #$00C014E2,a1
	beq    checka2
	bra    skip
checka2:
	cmpa.l #$FFFFFFFF,a2
	beq    checka3
	bra    skip
checka3:
	cmpa.l #$00FE86EE,a3
	beq    checka4
	bra    skip
checka4:
	cmpa.l #$00001558,a4
	beq    checka5
	bra    skip
checka5:
	cmpa.l #$00070000,a5
	beq    checka6
	bra    skip
checka6:
	cmpa.l #$00C00276,a6
	beq    checka7
	bra    skip
checka7:
	cmpa.l #$00C014B2,a7
	beq    success
	bra    skip
```

**Comment by mras0** [comment]
I'd missed that the disk also contains a custom bootblock that does naughty things before running startup-sequence. It loads a small program that assumes the upper bits of D5 are clear (see below).

Looks like KS1.3 doesn't initialize D5 or D7 before running the bootblock code (or rather it ends up preserving the initial value when it does).  Did you run your test from a normal executable? (If yes, perhaps you want to try this one instead: [showregs.zip](https://github.com/dirkwhoffmann/vAmiga/files/7498261/showregs.zip)).

WinUAE result:
  D0 00000000   D1 8000FFFF   D2 00000001   D3 00000000
  D4 00000000   D5 00000000   D6 FFFFFFFF   D7 00000000
  A0 00001958   A1 00C014E2   A2 FFFFFFFF   A3 00FE86EE
  A4 00001558   A5 00C014B6   A6 00C00276   A7 00C014B2

This matches what I get in my own emulator, and if I initialize all registers to some other value than 0 it only shows up in D5 and D7.
 
```
Start                           	; $0004e000
	MOVEQ.L	#$05, D6
	MOVE.W	#$0039, D7
lab_0004e006
	MOVEM.W	D6-D7, -(A7)
	MOVE.W	D6, D7
	MOVEQ.L	#$00, D4
	MOVE.W	D6, D4
	LSL.L	#$08, D4
	DIVU.W	D7, D4
	MULU.W	D4, D4
	MOVEA.L	D4, A0
	NEG.L	D4
	ADDA.L	A0, A0
	CLR.L	D0
	MOVE.W	D6, D0
	MOVE.L	D0, D1
	ADD.W	D1, D1
	SWAP.W	D1
	LEA.L	dat_00050000, A1
	MOVEA.L	A1, A3
	MOVEA.L	A1, A4
	ADDA.L	#$0000f000, A4
	MOVE.W	#$00f0, D2
	MOVE.W	D2, D5			; Only word sized
	MOVE.W	D2, D3
	SUBQ.W	#$01, D3
	LSL.W	#$07, D5		; Only word sized
	ADDA.L	D5, A1			; !!!!
```

**Comment by dirkwhoffmann** [comment]
I tried the demo on my A500 (Rev. 8A). What I did is to copy the ADF to a memory stick, plug the stick into the Gotek drive and switch the machine on. To my surprise, the demo didn’t boot. It hangs very early, showing a light grey screen. After doing a soft reset (Ctrl-Amiga-Amiga), I got a guru meditation:

![guru](https://user-images.githubusercontent.com/12561945/140817978-ff23f88e-ebba-4bce-ad16-65ab2f611779.jpg)

After another soft reset, it  hanged again, showing the light grey screen. The next reset produced the Guru and so on. Interestingly, this is different to both vAmiga and UAE.

After switching on the machine with showregs.adf inserted, I got the values shown to the left: 

![values](https://user-images.githubusercontent.com/12561945/140818495-68972957-7774-4b7a-8abb-af6ce1d2a62e.jpg)

After trying a second time, I got the values shown to the right (D5 differs).

We seem to have the following situation:
- On some machines (including the one of the developer), the initial bit value (for most bits) is 0.  
- On some machines (including mine), the initial bit value (for most bits) is 1.
- There is some randomness involved (on my machine: bit 15 in D5 sometimes toggles).

> EDIT: Just to be clear: I think setting the registers to 0 is best for compatibility even if they're formally undefined at startup (MC68000UM section 5.5 says only PC/SSP and IPL are touched on RESET). Programs that rely on specific values are obviously buggy, but apparently exist.

Oh, I've just seen your edit. Maybe, I'll add a (hidden power-user) option OPT_CPU_INIT_PATTERN because it would be a nice addition to the already existing OPT_RAM_INIT_PATTERN. All registers set to 0 should be the default setting.

**Comment by dirkwhoffmann** [comment]
@mras0: Do you know who the author of 'showregs' is? I want to copy the program to my test case repo and I'd like to give credit to the author in the README file. I know it's a tiny program, but still...

**Comment by mras0** [comment]
Very interesting results! Maybe they didn't test cold boot (enough) or other revisions behaved differently.

> @mras0: Do you know who the author of 'showregs' is? I want to copy the program to my test case repo and I'd like to give credit to the author in the README file. I know it's a tiny program, but still...

Yes, it's me :) Just something I hacked together quickly. Consider it public domain/MIT licensed/whatever, you're certainly free to use it whatever way you like. (Note: to reproduce the ADF you need to calculate the right checksum and place it at offset 4)

**Comment by dirkwhoffmann** [comment]
Very cool. I think this one can be considered resolved. vAmiga's got a new option `OPT_REG_RESET_VAL` that can be used to alter the initial CPU register value. By default, it is set to 0. It's not exposed to the GUI, but accessible via RetroShell as shown in the following regression testing script:

```
# Setup the test environment
regression setup A500_ECS_1MB /tmp/kick13.rom

# Change the configuration and restart 
cpu set regreset 0xFFFFFFFF
amiga reset

# Run the test
regression run /tmp/showregs2.adf
wait 9 seconds

# Exit with a screensht
screenshot set cutout 196 36 908 314
screenshot save showregs2
```

`showregs` is now part of the test repo (in `vAmigaTS/Init/cpu/`). I simply copied the ADF. It can be used directly as a regression test in combination with the test script above.

**Comment by mras0** [comment]
And by quickly hacking it, I ended up making many of the same mistakes as previous coders. Here's a fixed and slightly cleaned version, that you should probably include instead (to not tarnish my already spotty reputation too much more :)): 

[showregs-fixed.zip](https://github.com/dirkwhoffmann/vAmiga/files/7507027/showregs-fixed.zip)

Mistakes in the old version:

- Didn't initialize DIWSTRT/DIWSTOP/DDFSTRT/DDFSTOP/BPLxMOD -> Doesn't work with e.g. KS3.0
- Didn't actually disable interrupts/DMA like I meant to -> Crashes/misbehaves in other configs than 512K chip/512K slow
- Didn't correctly set the DMA configuration I wanted (DMAEN (bit 9) not set), only worked because of previous bug

Just shows how easy it is to create compatibility problems when it works on "your machine" I guess.

Results from trying a few different KS's (just with my own emulator, because I couldn't be bothered to modify WinUAE, so take it FWIW especially the KS1.0 results):

 - KS 1.0+KS1.0/KS1.1 disk: Garbled screen (I guess topaz.8 has different format)
 - KS 1.0+KS1.3 disk: Only D7 survives
 - KS 1.2/1.3: D5 & D7 survives
 - KS 1.4: Only D7 survives
 - Later: No registers survive

**Comment by dirkwhoffmann** [comment]
> Here's a fixed and slightly cleaned version, that you should probably include instead

Thanks. I've replaced the ADF and the source file in the test repo.

Also, I've peeked into your source code to see how text messages can be printed (since I never wrote an assembly program on the Amiga back in the day, I am a still newbie in this field). Having learned how to do this, I am able to improve some of my test cases quite a bit. Instead of displaying register values by color bars, I can now print the values directly. I'll update some of my test case accordingly...

<img width="1024" alt="Bildschirmfoto 2021-11-10 um 15 26 21" src="https://user-images.githubusercontent.com/12561945/141132446-49e6b6f1-6b6f-4ae1-81b7-87adee89b8c7.png">

**Comment by mithrendal** [comment]
@mras0 I like the colors and the font 🤤🤤🤤

**Comment by mras0** [comment]
> Thanks. I've replaced the ADF and the source file in the test repo.
> 
> Also, I've peeked into your source code to see how text messages can be printed (since I never wrote an assembly program on the Amiga back in the day, I am a still newbie in this field). Having learned how to do this, I am able to improve some of my test cases quite a bit. Instead of displaying register values by color bars, I can now print the values directly. I'll update some of my test case accordingly...

Cool! But be aware that it's still quick&dirty :) The code hardcodes values/knowledge that really should be read from the TextFont structure, e.g. tf_LoChar (why I subtract 32 from the character) and tf_Modulo (256-64). But these assumptions are "OK" for topaz.8 on KS1.2-3.1 at least, where you're also guaranteed that it's available and located in ROM. There's more info the [manuals](http://amigadev.elowar.com/read/ADCD_2.1/Libraries_Manual_guide/node03DE.html). For earlier versions the assumptions don't hold, so keep that in mind for the A1000 case (if/when you need to test KS < 1.2).

Speaking of which, I think it should be easy to create a replacement "test kickstart" disk that would load on such a system and run a bunch of tests very early. Or of course, if somebody has a system that allows (reasonably easy) replacement of the ROM, we could just create a new one (I created a basic "framework" for that a couple of years ago: https://github.com/mras0/simple-amiga-rom).

@mithrendal: Who could not love topaz.8? :)

**Comment by dirkwhoffmann** [comment]
> But be aware that it's still quick&dirty :) 

It's perfectly fine, it serves our needs. Just one little thing:
  
```
start:
        movem.l d0-a7,-(a7)     ; Store registers
        move.l  a6,a5           ; Execbase in a5
```

This only works when the disk is in the drive before resetting. If the disk is inserted after the hand&disk logo has shown up, a6 does not contain Execbase any more.

> Speaking of which, I think it should be easy to create a replacement "test kickstart" disk that would load on such a system and run a bunch of tests very early.

I've bought an ACA 500 plus a couple of month ago. I think I can upload any Kickstart I want via one of the memory card slots. I'll try this out with one of the DiagRoms when I'll find some more time...

**Comment by mras0** [comment]
> 
> 
> > But be aware that it's still quick&dirty :)
> 
> It's perfectly fine, it serves our needs. Just one little thing:
> 
> ```
> start:
>         movem.l d0-a7,-(a7)     ; Store registers
>         move.l  a6,a5           ; Execbase in a5
> ```
> 
> This only works when the disk is in the drive before resetting. If the disk is inserted after the hand&disk logo has shown up, a6 does not contain Execbase any more.
> 

Really? What configuration? Obviously easily fixed, but now I'm curious as I can't reproduce it.

[RKM Devices](http://amigadev.elowar.com/read/ADCD_2.1/Devices_Manual_guide/node007D.html) says "The boot code is invoked with [...] SysBase in A6." and e.g. [Amiga System Programmers Guide](https://archive.org/details/Amiga_System_Programmers_Guide_1988_Abacus/page/n359/mode/2up) (section 3.2.1 on page 352) also says: "A pointer to the Exec base is passed in A6"

So, I think it should be safe to assume that A6 contains Execbase (this matches my reading of KS1.2/KS1.3 disassembly).

As far as I can tell the boot block code is only called like this (from KS1.2, but 1.3 is very similar):

```
	; Read boot block to A4, A6 = Execbase
	MOVE.W	#CMD_READ, io_Command(A1)
	MOVE.L	#$00000400, io_Length(A1)
	MOVE.L	A4, io_Data(A1)
	MOVE.L	#$00000000, io_Offset(A1)
	JSR	_LVODoIO(A6)
	TST.L	D0
	BNE.B	lab_00fe8a5c
	MOVE.L	(A4), D0
	CMP.L	dat_00fe88c0(PC), D0	; dat_00fe88c0 -> 'DOS\0'
	BNE.B	lab_00fe8a5c
	; Calculate checksum
	MOVEA.L	A4, A0
	MOVE.W	#$00ff, D1
	MOVEQ.L	#$00, D0
lab_00fe8a1a
	ADD.L	(A0)+, D0
	BCC.B	lab_00fe8a20
	ADDQ.L	#$01, D0
lab_00fe8a20
	DBF.W	D1, lab_00fe8a1a
	NOT.L	D0
	BNE.B	lab_00fe8a5c
	; A1 = IORequest
	LEA.L	$002c(A5), A1
	; Call entry point
	JSR	$000c(A4)
```


> > Speaking of which, I think it should be easy to create a replacement "test kickstart" disk that would load on such a system and run a bunch of tests very early.
> 
> I've bought an ACA 500 plus a couple of month ago. I think I can upload any Kickstart I want via one of the memory card slots. I'll try this out with one of the DiagRoms when I'll find some more time...

Cool, absolutely no rush, it was just something I though of :)

**Comment by dirkwhoffmann** [comment]
> Really? What configuration? Obviously easily fixed, but now I'm curious as I can't reproduce it.

OK, I think I am wrong here. What I experienced is that the showregs program sometimes ended up in a black screen when the disk was inserted after the hand&disk logo had shown up. When the disk was inserted on boot time, it always worked. But it's hard to reproduce. With the (fixed) version, I had a black screen once today. After that, it always worked.

**Comment by mras0** [comment]
> 
> 
> > Really? What configuration? Obviously easily fixed, but now I'm curious as I can't reproduce it.
> 
> OK, I think I am wrong here. What I experienced is that the showregs program sometimes ended up in a black screen when the disk was inserted after the hand&disk logo had shown up. When the disk was inserted on boot time, it always worked. But it's hard to reproduce. With the (fixed) version, I had a black screen once today. After that, it always worked.

Definitely sounds like a bug. Are you seeing this on your real Amiga? Unfortunately I don't see an obvious cause this time. The screen should turn red on any detected error, it only changes to black just before enabling DMA and starting the infinite loop.

I'll try to see if I can make the code more robust and perhaps add some more color changes etc. that can help narrow it down.

**Comment by mras0** [comment]
Argh, another classic mistake. I disable DMA, setup customregs (including COP1LC) and re-enable DMA. If the last step happens in a place where the old copperlist is active it'll overwrite some of the custom registers. That's very likely the cause of the issue you're seeing (it's a bug for sure). Fix is to strobe COPJMP1 before re-enabling DMA.

[showregs-fixed2.zip](https://github.com/dirkwhoffmann/vAmiga/files/7528922/showregs-fixed2.zip)

(also includes a few minor improvements, but hopefully no new bugs)

**Comment by dirkwhoffmann** [comment]
@mras0: Thanks, I've updated the test repo with the new version. I didn't experience any issues with it so far.

**Comment by dirkwhoffmann** [comment]
As an experiment, I tried to flash the Logica Diagnostics Rom onto my ACA 500 plus. I found this tutorial on the web explaining the steps (in German, with a link to an English version): 

http://www.hirnwei.de/?p=1475

I followed the steps and managed to flash the Logica Rom onto the ACA 500 (via Kick 3.1 and WB 3.1). However, the flash program issued some warning messages about wrong checksums etc. Because it still finished the process, I tried to boot from the flashed image. At the beginning, it looked fine (Logica starts with some Copper bars). A few seconds later, it got stuck with a flickering power LED and those stripes on the screen:

![ ](https://user-images.githubusercontent.com/12561945/141687354-6cb497cd-efde-4d35-8c61-143ff0ff01df.JPG)

Either this is due to a flashing problem or the ACA 500 plus somehow interferes with the Rom. I don't know.
